morph_tool/morphdb.py

```diff
@@ -0,0 +1,346 @@
+    def from_folder(cls,
+                    morphology_folder: Path,
+                    mtypes: Dict[str, str],
+                    label: str = 'default'):
+        '''Factory method to create a MorphDB object from a folder containing morphologies
+
+        Args:
+            morphology_folder: a folder containing morphologies
+            mtype: a dictionary of where key is a morphology name and value its mtype
+            label: (optional) a group label to be used to identify the morphlogies from this folder
+        '''
+
+        obj = cls()
+        obj += (MorphInfo(path.stem, mtypes[path.stem], label=label)
+                for path in iter_morphology_files(morphology_folder))
+        return obj
+
+    def write(self, output_path: Path):
+        '''Write the neurondb file to XML or DAT format'''
+        ext = output_path.suffix.lower()
+        if ext == '.dat':
+            self.df[['name', 'layer', 'mtype']].to_csv(output_path, sep=' ', header=False,
+                                                       index=False)
+        elif ext == '.xml':
+            def get_repair_attr(morph, attr):
+                if attr == 'axon_inputs':
+                    return 'axon_sources', {'axoninput': getattr(morph, attr)}
+                else:
+                    return attr, getattr(morph, attr)
+            data = {'neurondb':
+                    {'listing': {
+                        'morphology': [{
+                            'name': morph.name,
+                            'mtype': morph.mtype_no_subtype,
+                            'msubtype': morph.msubtype,
+                            'layer': morph.layer,
+                            'repair': dict(
+                                get_repair_attr(morph, attr)
+                                for attr in BOOLEAN_REPAIR_ATTRS + ['axon_inputs']
+                            )
+                        } for morph in self.df.itertuples()]}
+                     }}
+            with output_path.open('w') as fd:
+                fd.write(xmltodict.unparse(data, pretty=True))
+        else:
+            raise ValueError(f'Unsupported neurondb extensions ({ext}).'
+                             ' Should be one of: (xml,csv,dat)')
+        return output_path
+
+    def features(self, config: Dict, n_workers=1):
+        '''Returns a dataframe containing morphometrics and neurondb information
+
+        Args:
+            config: a NeuroM morph_stas config.
+                See https://neurom.readthedocs.io/en/latest/morph_stats.html for more information
+
+        Returns:
+            A jointure dataframe between `neurom.stats.extract_dataframe` and `self.df`
+
+        Raises:
+            ValueError: if `self.df` has undefined (ie. None) paths
+        '''
+        missing_morphs = self.df[self.df.path.isnull()].name.values
+        if missing_morphs:
+            raise ValueError(
+                f'DataFrame has morphologies with undefined filepaths: {missing_morphs}')
+
```

Comment:
this was a bit broken for me witht multi-index, as stats df did not have multi-index. I did that:
```
        df = self.df.copy().reset_index(drop=True)                                                  
        df.columns = pd.MultiIndex.from_product((["neuron"], df.columns.values))                    
        stats = extract_dataframe(df['neuron', 'path'], config, n_workers).drop(columns='name')     
        stats.columns = pd.MultiIndex.from_product((["stats"], stats.columns.values))               
        return df.join(stats, how='inner')  
```

Reply:
I am stupid, I was not using latest Neurom version, all good!